modules/dnn/src/layers/recurrent2_layers.cpp

```diff
@@ -0,0 +1,611 @@
+
+            outTailShape.clear();
+        }
+
+        bool getMemoryShapes(const std::vector<MatShape> &inputs,
+                         const int requiredOutputs,
+                         std::vector<MatShape> &outputs,
+                         std::vector<MatShape> &internals) const CV_OVERRIDE
+        {
+            const MatShape& inp0 = inputs[0];
+            const MatShape& Wx = inputs[1];
+            const MatShape& Wh = inputs[2];
+
+            int _hidSize = Wh[2];
+            int _inpSize = Wx[2];
+            MatShape outTailShape_(outTailShape), outResShape;
+
+            if (!outTailShape_.empty())
+                CV_Assert(total(outTailShape_) == _hidSize);
+            else
+                outTailShape_.assign(1, _hidSize);
+
+            // compute output shape of y
+            // figure out batch size
+            int _batchSize;
+            int _seqLen;
+            if (useTimestampDim)
+            {
+                CV_Assert(inp0.size() >= 2 && total(inp0, 2) == _inpSize);
+                if (layout == SEQ_BATCH_HID) {
+                    _batchSize = inp0[1];
+                    _seqLen = inp0[0];
+                } else {
+                    _batchSize = inp0[0];
+                    _seqLen = inp0[1];
+                }
+                outResShape.push_back(_seqLen);
+            }
+            else
+            {
+                CV_Assert(inp0.size() >= 2 && total(inp0, 1) == _inpSize);
+                _batchSize = inp0[0];
+            }
+
+            outResShape.push_back(1 + static_cast<int>(bidirectional));
+            outResShape.push_back(_batchSize);
+            outResShape.push_back(_hidSize);
+            outputs.assign(1, outResShape);
+
+            int shp[] = {1 + static_cast<int>(bidirectional), _batchSize, numHidden};
+            MatShape newShape(shp, shp + sizeof(shp)/sizeof(shp[0]));
+
+            // compute output shape of yc
+            if (produceCellOutput)
+            {
+                outputs.push_back(newShape);
+            }
+            // compute output shape of yh
+            if (produceOutputYh)
+            {
+                outputs.push_back(newShape);
+            }
+
+            // internal shapes need during forward pass
+            internals.assign(1, shape(_batchSize, _hidSize)); // hInternal
+            internals.push_back(shape(_batchSize, _hidSize)); // cInternal
+            internals.push_back(shape(_batchSize, 1)); // dummyOnes
+            internals.push_back(shape(_batchSize, 4*_hidSize)); // gates
+
+            return false;
+        }
+
+        void getTypes(const std::vector<MatType>& inputs,
+                      const int requiredOutputs,
+                      const int requiredInternals,
+                      std::vector<MatType>& outputs,
+                      std::vector<MatType>& internals) const CV_OVERRIDE
+        {
+            outputs.assign(requiredOutputs, inputs[0]);
+            internals.assign(4, inputs[0]);
```

Comment:
According to implementation, only fp32 and m.b. fp64 are supported. Need ti add check/assert here.